src/server/scripts/Kalimdor/RagefireChasm/boss_slagmaw.cpp

```diff
@@ -0,0 +1,160 @@
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "Containers.h"
+#include "InstanceScript.h"
+#include "ScriptMgr.h"
+#include "ScriptedCreature.h"
+#include "ragefire_chasm.h"
+
+enum SlagmawSpells
+{
+    SPELL_LAVA_SPIT             = 119434,
+    SPELL_MAGNAW_SUBMERGE       = 120384,
+    SPELL_MAGNAW_TELEPORT_NORTH = 119424, // Serverside
+    SPELL_MAGNAW_TELEPORT_EAST  = 119425, // Serverside
+    SPELL_MAGNAW_TELEPORT_SOUTH = 119426, // Serverside
+    SPELL_MAGNAW_TELEPORT_WEST  = 119428  // Serverside
+};
+
+enum SlagmawEvents
+{
+    EVENT_LAVA_SPIT       = 1,
+    EVENT_TELEPORT        = 2,
+    EVENT_EMERGE          = 3
+};
+
+std::array<uint32, 4> const TeleportSpells =
+{
+    SPELL_MAGNAW_TELEPORT_NORTH, SPELL_MAGNAW_TELEPORT_EAST, SPELL_MAGNAW_TELEPORT_SOUTH, SPELL_MAGNAW_TELEPORT_WEST
+};
+
+// 61463 - Slagmaw
+struct boss_slagmaw : public BossAI
+{
+    boss_slagmaw(Creature* creature) : BossAI(creature, BOSS_SLAGMAW), _lavaSpitCounter(0), _lastTeleportSpell(0) { }
+
+    void JustAppeared() override
+    {
+        me->SetCanMelee(false);
+    }
+
+    void Reset() override
+    {
+        _Reset();
+        _lavaSpitCounter = 0;
+        _lastTeleportSpell = 0;
+    }
+
+    void JustDied(Unit* /*killer*/) override
+    {
+        _JustDied();
+        instance->SendEncounterUnit(ENCOUNTER_FRAME_DISENGAGE, me);
+    }
+
+    void EnterEvadeMode(EvadeReason /*why*/) override
+    {
+        _EnterEvadeMode();
+        _DespawnAtEvade();
+
+        instance->SendEncounterUnit(ENCOUNTER_FRAME_DISENGAGE, me);
+    }
+
+    void JustEngagedWith(Unit* who) override
+    {
+        BossAI::JustEngagedWith(who);
+        instance->SendEncounterUnit(ENCOUNTER_FRAME_ENGAGE, me, 1);
+        events.ScheduleEvent(EVENT_LAVA_SPIT, 1s);
+    }
+
+    void PhaseSubmerge()
+    {
+        DoCastSelf(SPELL_MAGNAW_SUBMERGE);
+        _lavaSpitCounter = 0;
+
+        events.ScheduleEvent(EVENT_TELEPORT, 3s);
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
+
+        events.Update(diff);
+
+        if (me->HasUnitState(UNIT_STATE_CASTING))
+            return;
+
+        if (me->GetVictim() && me->GetVictim()->GetDistance(me) > 50.0f)
+            EnterEvadeMode(EvadeReason::Other);
+
+        switch (events.ExecuteEvent())
+        {
+            case EVENT_LAVA_SPIT:
+            {
+                if (_lavaSpitCounter < 5)
+                {
+                    if (Unit* target = SelectTarget(SelectTargetMethod::Random, 0))
+                    {
+                        DoCast(target, SPELL_LAVA_SPIT);
+                        _lavaSpitCounter++;
+                    }
+                    events.Repeat(1s);
+                    break;
+                }
+                else if (_lavaSpitCounter == 5)
+                {
+                    PhaseSubmerge();
+                    events.CancelEvent(EVENT_LAVA_SPIT);
```

Comment:
no need to cancel it here, as its not re-scheduled in this `else if` branch, its only rescheduled in `if` above